Feature/suin

whl5105 opened the pull request:
### 데모 영상
![화면 기록 2023-03-17 오전 10 44 55](https://user-images.githubusercontent.com/73993670/225791356-9f0e78bb-8557-4a47-ae44-4395fd90da3c.gif)




## 요구사항
1. 시계열 차트 만들기
    - [x] 주어진 JSON 데이터의 `key`값(시간)을 기반으로 시계열 차트를 만들어주세요
    - [x] 하나의 차트안에 Area 형태의 그래프와 Bar 형태의 그래프가 모두 존재하는 복합 그래프로 만들어주세요
    - [x] Area 그래프의 기준값은 `value_area` 값을 이용해주세요
    - [x] Bar 그래프의 기준값은 `value_bar` 값을 이용해주세요
    - [x] 차트의 Y축에 대략적인 수치를 표현해주세요(예시 이미지 참고)
2. 호버 기능 구현
    - [x] 특정 데이터 구역에 마우스 호버시 `id, value_area, value_bar` 데이터를 툴팁 형태로 제공해주세요
3. 필터링 기능 구현
    - [x] 필터링 기능을 구현해주세요, 필터링은 특정 데이터를 하이라이트 하는 방식으로 구현해주세요
    - [x] 필터링 기능은 버튼 형태로 ID값(지역이름)을 이용해주세요
    - [x] 필터링 시 버튼에서 선택한 ID값과 동일한 ID값을 가진 데이터 구역만 하이라이트 처리를 해주세요
    - [x] 특정 데이터 구역을 클릭 시에도 필터링 기능과 동일한 형태로 동일한 ID값을 가진 데이터 구역을 하이라이트해주세요

## 구현 방법

> **query string을 활용하여 지역구 필터링 구현**
- '/' 경로 진입 또는 id="" 빈값으로 필터링할 경우 페이지 Navigate 처리
- useSearchParamsState hook을 사용하여 분리된 컴포넌트에서 hook 재사용

<br />

> **chartJs-2를 사용하여 시계열 차트 구현**
- `하이라이트` : ParamsState 값과 데이터를 비교하여 하이라이트 처리
     ('area' - pointBorderColor , segmentColor    , 'bar'- backgroundColor )
- `tooltip` : 스타일링 커스텀 및 id 데이터 추가  

<br />

## 개선 및 수정 사항 
1. **페이지 진입시 불필요한 페이지 이동처리 삭제** 
/ 경로에서 쿼리 스트링값을 전체로 넣으려고 불필요한 페이지 이동처리를 한것같다고 생각되서 
?id= "전체" 값을 별도로 선언하지 않아도 전체 데이터가 나올 수 있게 코드를 변경했습니다. 
2. **필터 드롭다운 -> 버튼 수정** 
 준수님이 작업해주신 버튼에서 컬러값들을 조금 수정해서 적용해보았습니다 ! 
3. ** 전역 스타일링 적용 ** 
4. ** area 차트 segment 별 하이라이트 추가 구현 **
기존에 포인트 부분만 하이라이트 처리를 했었는데 추가로 각 segment의 값까지 추가로 변경하였습니다 !  
5. ** 차트데이터 hook 파일 분리 ** 


<br />

whl5105 committed:
[#1] Init : mui , chartjs 패키지 추가

- 차트 구현 chartjs 설치
- 차트 구역 필터링 mui 설치

whl5105 committed:
[#2] Docs : 전역 타입 선언

Response data 타입을 여러컴포넌트에서 사용하고 있으므로 전역에 추가 해주었습니다

whl5105 committed:
[#2] Feat : axios api 선언

whl5105 committed:
[#2] Feat :  react router  연결 및 query string 리다이렉트 처리

QueryParmRoute 에서 query string 조건을 확인하며메인페이지 이동 처리를 해주었습니다.

whl5105 committed:
[#2] Docs : index.html 불필요한 주석 삭제  및 lang , noscript 수정

- html lang :ko 한글 적용
- noscript 한글 문구 추가
- 불필요한 주석 삭제

whl5105 committed:
[#2] Docs : 로고 이미지 추가

- header 영역에 들어가는 로고 이미지 추가

whl5105 committed:
[#2] Feat : 시계열 차트 구현

NyeongDev commented:
- useSearchParams 훅을 활용해서 필터링를 구현하신 점이 좋았습니다!
- 잘못된 경로로 진입했을 경우를 고려해서 router에 errorelemnt 넣으신 것도 좋았어요!
- 데이터를 하이라이트 할 때 선택되지 않은 데이터 색을 연하게 바꾸는 것이 사용자 입장에서 보기에 자연스러워서 좋은 방법인 것 같습니다!
- 추가 개선사항에 써 주신 것처럼 x 축 스크롤링을 구현하는 건 정말 좋은 아이디어 같아요! 저도 x 축 데이터가 많아서 확대해서 살펴보기가 불편하다고 생각하고 있었습니다.
- dropDown으로 필터링을 구현하신 것도 저는 좋은 방법인 것 같은데 요구 사항에서 필터링 기능은 버튼 형태로 구현해달라고 했던 것 같습니다..!

수고하셨습니다👍

whl5105 commented:
오호! 제가 오해를 했군요 버튼으로 수정해보겠습니다! 감사해요☺️

eunrain commented:
수인님 차트 역시나 예쁩니다..! 깨알 Footer까지..👍
저도 useSearchParams 훅 사용해서 구현하신 점 정말 인상 깊게 보았습니다!
드롭다운도 간단하게 구현할 수 있었다니.. mui 하나 배워갑니다!! 
고생하셨어요!

whl5105 committed:
[#2] Docs : 타입명 수정

whl5105 committed:
[#2] Docs : compilerOptions target 'es5' -> 'es6' 변경

whl5105 committed:
[#2] Docs : App.tsx (layout component) 파일 삭제

- header footer 없이 route Outlet 적용

whl5105 committed:
[#2] Feat : 차트 area 하이라이트 기능 추가

포인터를 기준으로 선택 시 다음 포인터 영역 까지  배경 하이라이트 추가

whl5105 committed:
[#2] Style : 배경컬러  변경

기본 컬러 -> 그라데이션 적용

whl5105 committed:
[#2] Feat : 필터 드롭다운 ->  버튼 구현 변경 , click 이벤트 함수 위치 수정

- onClick 시 불필요한 추가 함수 선언 정리
-  필터 드롭다운(mui) ->  버튼 구현 변경(스타일 컴포넌트)

whl5105 committed:
[#2] Docs : App.tsx (layout component) 파일 삭제

-  This reverts commit 6083c150c7fc74a3f35bea380498a2d4c99ad1ba.

eunrain commented on src/hooks/useParams.ts:
```diff
@@ -0,0 +1,21 @@
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const searchParam = searchParams.get(searchParamName);
+  const searchParamsState: string = searchParam ?? defaultValue;
```
수인님! defaultValue는 ID값(지역이름)이 없을때, 즉 전체 data를 보여주기 위해 추가해주신 부분일까요??

whl5105 replied:
네! 맞습니다 기본값을 전체로 설정해도 되지만 재사용성을 고려해서 defaultValue 가 필요한 부분은 인자로 받게 구현했습니다

whl5105 committed:
[#2] Style : 전역 스타일  theme 추가

- styled-components 의 ThemeProvider , GlobalStyle 적용

ThemeProvider : 전역에서 사용하는 컬러 적용
GlobalStyle : 전역 스타일 초기값 선언

whl5105 committed:
[#2] Docs : useSearchParams 파일명 수정

whl5105 committed:
[#2] Style : 메인 스타일 변경 , 코드 분리

차트 데이터 커스텀 훅 분리 및 페이지 스타일 변경

whl5105 committed:
[#2] Docs : 버튼 컴포넌트 위치 변경

- 공통 element 경우 ui 폴더로 파일 이동

whl5105 committed:
Revert "[#2] Docs : useSearchParams 파일명 수정"

This reverts commit 5fbd9ba45c109d15a013dc3a340eb02ffbc462d2.

whl5105 committed:
[#2] Docs : useSearchParams 파일명 수정

whl5105 committed:
[#2] Docs,Feat: 타임차트 파일명 수정  , 필터 및 타임차트 코드 개선

NyeongDev commented:
ThemeProvider를 적용해서 색상을 재사용 할 수 있도록 하신 점이 좋았습니다! area차트에도 하이라이트 기능 적용하셔서 차트가 눈에 확 들어오는 것 같아요! 수고하셨습니다~

whl5105 committed:
[#2] Feat : 페이지 진입시 Navigate 처리 삭제

페이지 진입시 페이지 이동 처리 삭제
-기본 path="/"  전체로 적용

yeon-kk commented on src/hooks/useChartDataState.ts:
```diff
@@ -0,0 +1,36 @@
+      .finally(() => setIsLoading(props => !props));
+  }, []);
+
+  const dateValues: string[] = Object.keys(resData);
```
useCharDataState로 따로 분리하면서 Object 메서드를 keys와 values로 수정하신 것으로 기억하는데 이 방법이 리턴하는 값을 더 명확하게 표현한 것 같아요

eunrain commented:
관심사를 분리해서 코드가 훨씬 간결하고, 가독성이 좋아진 것 같습니다! 차트도 깔끔하고 아주 멋드러집니다!!😎 고생하셨어요 !

yeon-kk reviewed (commented):
주석을 없애고 코드를 같이 수정해주셨는데 더 읽기 편해진 것 같아요~! area 데이터 구역에 대한 필터링 구현해내셨군요 멋있습니다!

whl5105 committed:
[#2] Style : 불필요한 스타일 선언 삭제 , 스타일 코드 변경

- 버튼에 사용되는 box-shadow 삭제
- 메인 페이지 위치값 수정
- theme rgba 컬러 변경

whl5105 committed:
[#2] Refactor : 페이지 접근시 NotFound -> errorElement Redirect 처리 변경

프로젝트가 단일 페이지 동작으로 인해  NotFound 보다는 메인 페이지로 이동하는 방법으로 변경 되었습니다.

whl5105 committed:
[#2] Refactor : useSearchParams 선언방식 변경 , 데이터 처리 개선

- 기존 useSearchParams hook보다는 각 컴포넌트 에서 선언하여 적용하는 방식으로 변경
-"전체" 버튼 은 별도로 button 컴포넌트 사용하는 방식으로 변경
- 쿼리 스트링 값을 반영하여 하이라이트 컬러 변경 함수 수정

whl5105 committed:
[#2] Refactor : 불필요한 파일 , 패키지 삭제

whl5105 committed:
[#2] Refactor : public 파일 정리
